pypesto/objective/aesara.py

```diff
@@ -5,139 +5,235 @@
+        return mode == MODE_FUN
+
+    def check_sensi_orders(self, sensi_orders, mode) -> bool:
+        if not self.check_mode(mode):
+            return False
+        else:
+            return self.base_objective.check_sensi_orders(sensi_orders, mode)
+
+    def call_unprocessed(
+            self,
+            x: np.ndarray,
+            sensi_orders: Tuple[int, ...],
+            mode: str,
+            **kwargs
+    ) -> ResultDict:
+
+        # hess computation in aesara requires grad
+        if 2 in sensi_orders and 1 not in sensi_orders:
+            sensi_orders = (1, *sensi_orders)
+
+        # this computes all the results from the inner objective, rendering
+        # them accessible to aesara compiled functions
+        self.inner_ret = self.base_objective.call_unprocessed(
+            self.infun(x), sensi_orders, mode, **kwargs
+        )
+        ret = {}
+        if 0 in sensi_orders:
+            ret[FVAL] = float(self.afun(x))
+        if 1 in sensi_orders:
+            ret[GRAD] = self.agrad(x)[0]
+        if 2 in sensi_orders:
+            ret[HESS] = self.ahess(x)[0]
+
+        return ret
+
+    def __deepcopy__(self, memodict=None):
+        other = AesaraObjective(
+            copy.deepcopy(self.base_objective), self.aet_x, self.aet_fun,
+            self._coeff
+        )
+
+        return other
+
+
+class AesaraObjectiveOp(Op):
     """
     Aesara wrapper around a (non-normalized) log-probability function.
 
     Parameters
     ----------
-    problem:
-        The `pypesto.Problem` to analyze.
-    beta:
-        Inverse temperature (e.g. in parallel tempering).
+    obj:
+        Base aseara objective
+    coeff:
+        multiplicative coefficient for the objective function value
     """
 
     itypes = [aet.dvector]  # expects a vector of parameter values when called
     otypes = [aet.dscalar]  # outputs a single scalar value (the log prob)
 
-    def __init__(self, problem: Problem, beta: float = 1.):
-        self._objective: Objective = problem.objective
-
-        def log_prob(x):
-            return - beta * self._objective(x, sensi_orders=(0,))
-
-        # initialize the log probability Op
-        self._log_prob = log_prob
+    def __init__(self,
+                 obj: AesaraObjective,
+                 coeff: Optional[float] = 1.):
+        self._objective: AesaraObjective = obj
+        self._coeff: float = coeff
 
         # initialize the sensitivity Op
-        if problem.objective.has_grad:
-            self._log_prob_grad = AesaraLogProbabilityGradient(problem, beta)
+        if obj.has_grad:
+            self._log_prob_grad = AesaraObjectiveGradOp(obj, coeff)
         else:
             self._log_prob_grad = None
 
     def perform(self, node, inputs, outputs, params=None):
-        theta, = inputs
-        log_prob = self._log_prob(theta)
+        log_prob = self._coeff * self._objective.inner_ret[FVAL]
         outputs[0][0] = np.array(log_prob)
 
     def grad(self, inputs, g):
```

Comment:
~~how~~by what dark magic is it possible that inputs is not used here? :thinking:

Reply:
Added a note about this. The trick is that we already know what the inputs here are going to look like when computing `inner_ret`, so we can ignore them here and just use the precomputed values. This also has the advantage that we only spend a minimal amount of time in aesara functions, since the bulk of computation is done in the outer `call_unprocessed`, which improves performance in multithreaded settings as we avoid a lot of locking of aesara functions.